### PR TITLE
entensions/filters/http/ext_authz: Added connection level metadata

### LINKED ISSUE #25762: envoy.filters.http.ext_authz: ability to pass connection metadata in authz request
*Title*: *Add ability to pass connection metadata in authz request*

*Description*:
External Authorization http extension currently supports passing dynamic filter metadata related to http request. I would like to propose a capability to pass downstream connection dynamic metadata typically added by transport socket extensions.

A particular problem I'm trying to solve is propagation of `envoy.transport_socket.peer_information/peer_identity` from [extensions.transport_sockets.alts.v3.Alts ](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/alts/v3/alts.proto.html#extensions-transport-sockets-alts-v3-alts) to the authorization checks.

I see 2 options: 
1. In [ext_authz.cc](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_authz/ext_authz.cc#L46-L66) walk through both `decoder_callbacks_->streamInfo().dynamicMetadata()` and `decoder_callbacks_->connection()->streamInfo().dynamicMetadata()` to populate `filter_metadata` and `typed_filter_metadata` fields in authz request. Request metadata should take precedence in an *unlikely case* when the same namespace is present in both connection metadata and requests metadata. Optionally this behaviour can be only enabled using an extra flag in configuration.

1. Add `connection_metadata_context_namespaces` and `connection_typed_metadata_context_namespaces` fields to [extensions.filters.http.ext_authz.v3.ExtAuthz proto](https://github.com/envoyproxy/envoy/blob/bfa849bc598870033dbbc11ae17c9abb273e19e5/api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto#L87-L107); and add `connection_metadata_context` field to the [auth attribute_context proto](https://github.com/envoyproxy/envoy/blob/42f82c257a049589b5c15788c774342268124921/api/envoy/service/auth/v2/attribute_context.proto#L160).

Happy to implement it myself after discussing the options above.

*Relevant Links*:
- `metadata_context_namespaces` in https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz

### PR DESCRIPTION
**Pull request by derpferd**
Commit Message: Added connection level metadata. Fixes #25762
Additional Description: Took the approach outlined as option 1 in #25762 
Risk Level: low
Testing: Unit tests added and manually ran local end to end test.
Docs Changes: Updated proto comment
Release Notes: Updated
Platform Specific Features:N/A

### CODE REVIEW

File: changelogs/current.yaml
Comment (phlax) [inline changelogs/current.yaml:92]:
```suggestion
    from connection metadata to external auth service. This is in addition to the current behavior of forwarding request metadata.
```

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:67]: ditto

**Review by tyxia — CHANGES_REQUESTED** [review]
Thanks for your contribution!  

I agree with high level approach that leverage the connection level dynamic metadata for your use case since that is where the data populated in ATLS.

I wanted to confirm the data flow of your use case is ATLS metadata-> ext_auth filter (here) -> ext_authz service? If so, just curious have you tested in a integrated test such usage.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (derpferd) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Done

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (derpferd) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Done

**Comment by derpferd** [comment]
`I wanted to confirm the data flow of your use case is ATLS metadata-> ext_auth filter -> ext_authz service?` - yes, we did a manual test which verified it worked

**Comment by tyxia** [comment]
LGTM, thanks

@derpferd , you will need to do a merge main and resolve the conflict

**Review by tyxia — APPROVED** [review]

**Comment by tyxia** [comment]
OR @ggreenway,  Could you take a look as code owner?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (wbpcode) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:62]: Prefer do this by a  `else if` then we needn't do the redundant searching and assignment if the key is exist in the request metadata.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (wbpcode) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:78]: ditto.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your contribution. Seems a reasonable feature and only one comment is added. Thanks again.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (derpferd) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Sorry for missing this the first time around

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (wbpcode) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:58]: I think we should try find request_metadata first because it has higher priority? Please add a test to ensure the request metadata will  be taken as priority if both are exist, Thanks.

And by the way, value semantic should be used for iterator, so `const auto metadata_it = ...` may be better here.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (wbpcode) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:75]: ditto.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for you quick update. Some new comments are added.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (derpferd) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:58]: You are right, the order is incorrect. This is why the presubmit tests are failing. I added the unit tests at: https://github.com/envoyproxy/envoy/pull/27612/files#diff-64a51bc8f4c7ce4f31a3420225e6fd4d0f14a187424148da4b9dc87e11d97c96R1426-R1432

**Review by tyxia — APPROVED** [review]
LGTM

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by tyxia** [comment]
@lizan  Could you please take a look as API Shepherd. Actually only comments are changed in API proto.

@derpferd  Sorry, could you merge main again as it has conflicts.

**Comment by tyxia** [comment]
@wbpcode or anyone else from @envoyproxy/api-shepherds  could take a look? The only things changed in API proto is comment.

I think it would be we can merge this PR, otherwise @derpferd will need to merge main again and agin

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (adisuissa) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:54]: why was the reference removed?

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:89]: Please add links to the relevant metadata fields that could be forwarded (I'm assuming one is for route, and the other is for listener, but not sure).
This should be clarified in the line above ("list of metadata").

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:104]: Same as above.

**Review by adisuissa — COMMENTED** [review]
I'm not expecting any conflicts, as the keys for the metadata fields should be different for the "request" and "connection" metadata entries. If this assumption isn't valid, please consider option 2 from #25762.

Left a couple of nits otherwise.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (derpferd) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:54]: This was done at the suggestion of wbpcode. https://github.com/envoyproxy/envoy/pull/27612#discussion_r1227462864

I don't have the needed experience with this to pro/con the usage of one or the other.

**Comment by tyxia** [comment]
> #25762

Here are my thoughts:
1. Technically, the key can conflict. It is just namespace and nothing in the code (e.g., check etc) to prevent the extension from setting both metadata using same key. But I agree that best practice probably is don't conflict like what ext_authz does.
2. The request overrides connection logic in this PR is under the circumstance that there is key conflicts. My understanding is that a connection can have many requests. Low level request metadata override high level connection metadata like the pattern of per route config -- router level overrides virtual host level

Regarding (1) above, I think we can suggest the best practice but i am not sure we can force extension to use different keys. 

Also, I don't have objection about option2 to simplify the logic by adding namespace for connection metadata specifically

**Comment by jmarantz** [comment]
I think Adi's last comment puts this in the author's hands for the next step.

/wait

**Comment by derpferd** [comment]
Created #28161 which implements the same feature except taking the safer option 2 approach.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (adisuissa) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:89]: I think this comment can be removed.

**Review by adisuissa — COMMENTED** [review]
@ggreenway this is the PR that we discussed offline.
Can you PTAL?
/assign @ggreenway

**Comment by adisuissa** [comment]
Ugh, repokitteh didn't assign @ggreenway. Assigning manually.

**Review by ggreenway — APPROVED** [review]
LGTM, aside from @adisuissa 's comment

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:89]: hmm, I am not sure why this can be removed.  There is no code change between recent approvals.  In the implementation, the request value takes precedence over connection value. 

As I mentioned before. It is just namespace and i am not sure we can force extension to use different keys. Low level request metadata override high level connection metadata is similar to the pattern of per route config that router level overrides virtual host level.

We probably can add comment to suggest the best practice of avoiding key conflicts.

**Comment by derpferd** [comment]
@ggreenway resolved @adisuissa 's comments so need another LGTM.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (adisuissa) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto]: You are correct that it is just a namespace and essentially can have any value.
However, the namespace should be well defined (and unique) by the filter that created it (probably [this doc](https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata) should be clarified, and some tests to ensure well-defined namespaces are used can be added).

**Comment by derpferd** [comment]
Merged from upstream, hoping this fixes failing coverage since the failing tests were unrelated to this change.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto]: Thanks for the links. Yea I agree it will be great to be clarified in the doc around namespace. such as unique and well defined namespace is strongly encouraged(opposite probably should be disallowed)

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by adisuissa** [comment]
@ggreenway approved, merging.

**Pull request merged**
